ComponentLib/PowerElectronicsComponents/SynchronousMachine/SynchronousMachine.cpp

```diff
@@ -0,0 +1,152 @@
+
+
+
+#include <iostream>
+#include <cmath>
+#include <vector>
+#include "SynchronousMachine.hpp"
+
+
+namespace ModelLib {
+
+
+
+/*!
+ * @brief Constructor for a constant SynchronousMachine model
+ *
+ * Calls default ModelEvaluatorImpl constructor.
+ */
+
+template <class ScalarT, typename IdxT>
+SynchronousMachine<ScalarT, IdxT>::SynchronousMachine(IdxT id, ScalarT Lls, std::tuple<ScalarT,ScalarT> Llkq, ScalarT Llfd, ScalarT Llkd, ScalarT Lmq, ScalarT Lmd, ScalarT Rs, std::tuple<ScalarT,ScalarT> Rkq, ScalarT Rfd, ScalarT Rkd, ScalarT J, ScalarT P, ScalarT mub)
+  : Lls_(Lls),
+    Llkq_(Llkq),
+    Llfd_(Llfd),
+    Llkd_(Llkd),
+    Lmq_(Lmq),
+    Lmd_(Lmd),
+    Rs_(Rs),
+    Rkq_(Rkq),
+    Rfd_(Rfd),
+    Rkd_(Rkd),
+    J_(J),
+    P_(P),
+    mub_(mub)
+{
+    this->size_ = 13;
+    this->n_intern_ = 6;
+    this->n_extern_ = 7;
+    this->extern_indices_ = {0,1,2,3,4};
+    this->idc_ = id;
+}
+
+template <class ScalarT, typename IdxT>
+SynchronousMachine<ScalarT, IdxT>::~SynchronousMachine()
+{
+}
+
+/*!
+ * @brief allocate method computes sparsity pattern of the Jacobian.
+ */
+template <class ScalarT, typename IdxT>
+int SynchronousMachine<ScalarT, IdxT>::allocate()
+{
+    this->y_.resize(this->size_);
+    this->yp_.resize(this->size_);
+    this->f_.resize(this->size_);
+    return 0;
+}
+
+/**
+ * Initialization of the grid model
+ */
+template <class ScalarT, typename IdxT>
+int SynchronousMachine<ScalarT, IdxT>::initialize()
+{
+    return 0;
+}
+
+/*
+ * \brief Identify differential variables
+ */
+template <class ScalarT, typename IdxT>
+int SynchronousMachine<ScalarT, IdxT>::tagDifferentiable()
+{
+    return 0;
+}
+
+/**
+ * @brief Contributes to the bus residual.
+ *
+ * Must be connected to a PQ bus.
+ * @todo fix up
+ */
+template <class ScalarT, typename IdxT>
+int SynchronousMachine<ScalarT, IdxT>::evaluateResidual()
+{
+    ScalarT rkq1 = std::get<0>(Rkq_);
+    ScalarT rkq2 = std::get<1>(Rkq_);
+    ScalarT llkq1 = std::get<0>(Llkq_);
+    ScalarT llkq2 = std::get<1>(Llkq_);
+
+    ScalarT cos1 = cos((P_/2.0)*y_[5]);
+    ScalarT sin1 = sin((P_/2.0)*y_[5]);
+    ScalarT cos23m = cos((P_/2.0)*y_[5] - (2.0/3.0)*M_PI);
+    ScalarT sin23m = sin((P_/2.0)*y_[5] - (2.0/3.0)*M_PI);
+    ScalarT cos23p = cos((P_/2.0)*y_[5] + (2.0/3.0)*M_PI);
+    ScalarT sin23p = sin((P_/2.0)*y_[5] + (2.0/3.0)*M_PI);
+
+    this->f_[0] = y_[6]*cos1 + y_[7]*sin1 + y_[8];
+    this->f_[1] = y_[6]*cos23m + y_[7]*sin23m + y_[8];
+    this->f_[2] = y_[6]*cos23p + y_[7]*sin23p + y_[8];
```

Comment:
This was embedded in the model as given. I suspect in the future something like that would happen again. Do we want to actively remove embedded Park's?

Reply:
It is the case-to-case decision, I guess. In this instance, I think it is better to have embedded Park, because all dq0 variables are used only internally in the machine.